### PR TITLE
Event bus and websocket endpoint

### PR DESCRIPTION
**Pull request by edvald**
Added an event bus on `garden.events`, and just to kick things off I'm emitting some events from the TaskGraph.

The WebSocket endpoint forwards any event emitted to `garden.events` to open
connections, and allows commands to be executed via the socket as well.

### CODE REVIEW

**Comment by eysi09** [comment]
Could you add idiot proof instructions for how to interact with the API endpoints? E.g. the shape of the JSON payload and the `command.subCommand` syntax.

File: garden-service/src/util/util.ts [hunk 1]
```diff
@@ -249,6 +250,38 @@ export function splitFirst(s: string, delimiter: string) {
   return [parts[0], parts.slice(1).join(delimiter)]
 }
 
+/**
+ * Recursively process all values in the given input,
+ * walking through all object keys _and array items_.
+ */
+export function deepMap<T extends object, U extends object = T>(
```
Comment (eysi09) [inline garden-service/src/util/util.ts]: This doesn't appear to be used anywhere.

File: garden-service/src/commands/build.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ import { Module } from "../types/module"
 import { logHeader } from "../logger/util"
 
 const buildArguments = {
-  module: new StringsParameter({
+  modules: new StringsParameter({
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/commands/build.ts]: Added a commit for that.

File: garden-service/src/commands/call.ts [hunk 3]
```diff
@@ -24,7 +24,7 @@ import dedent = require("dedent")
 
 const callArgs = {
   serviceAndPath: new StringParameter({
-    help: "The name(s) of the service(s) to call followed by the ingress path (e.g. my-container/somepath).",
+    help: "The name(s) of the service to call followed by the ingress path (e.g. my-container/somepath).",
```
Comment (eysi09) [inline garden-service/src/commands/call.ts]: Why remove the plural `(s)` from `service(s)`?

Comment (edvald) [reply garden-service/src/commands/call.ts]: We can't call more than one service here.

Comment (eysi09) [reply garden-service/src/commands/call.ts]: Ah yes. So `name` should also be singular than.

Comment (eysi09) [reply garden-service/src/commands/call.ts]: ...I'll go ahead and fix it and merge.

**Pull request merged**